### PR TITLE
Add cpp_std option for Visual Studio C++ compiler

### PR DESCRIPTION
**Pull request by polarina**

### CODE REVIEW

File: mesonbuild/compilers/cpp.py [hunk 1]
```diff
@@ -188,6 +188,9 @@ def get_options(self):
                                                    'C++ exception handling type.',
                                                    ['none', 'a', 's', 'sc'],
                                                    'sc'),
+                'cpp_std': coredata.UserComboOption('cpp_std', 'C++ language standard to use',
+                                                   ['none', 'c++11', 'c++14', 'c++17', 'c++latest'],
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/cpp.py]: __[Flake8]__

[E128] continuation line under-indented for visual indent

File: mesonbuild/compilers/cpp.py [hunk 2]
```diff
@@ -188,6 +188,9 @@ def get_options(self):
...
+                                                   'none'),
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/cpp.py]: __[Flake8]__

[E128] continuation line under-indented for visual indent

**Comment by jibsen** [comment]
Wouldn't you have to check the MSVC version? in VS2015 there is only c++14 and c++latest, and in versions before that I don't think the switch is there.

**Comment by polarina** [comment]
@jibsen right, I missed that. From what I gathered, MSVC tends to ignore unknown compiler flags, which is presumably why the VS2015 AppVeyor build surprisingly passed (VS2015 does not support C++17).

Speaking of which, is there a way to disable the new `13 vc c++17` test for VS2010 and VS2015 builds?

**Comment by nirbheek** [comment]
You can skip a test with `error('MESON_SKIP_TEST <msg>')`

**Comment by polarina** [comment]
The patches should be ready. I am unsure as to why the VS2015 build on AppVeyor is failing. If someone more knowledgeful about AppVeyour could take a look, that'd be appreciated.

Thanks!

**Comment by nirbheek** [comment]
@jpakkane could you please restart that appveyor job? It error out due to an unrelated problem.

**Comment by jpakkane** [comment]
The conceptual problem that this brings is that there are many projects that have default values for `cpp_std` set to, say, c++14. The question is what should we do if compiling these with VS? Currently there is no option so they are ignored. But what should happen in the future? Erroring out due to an invalid value is the simple thing but it's not a particularly good UX. Should they be remapped to "at least as new or newer version" so that for example c++11 gets remapped to c++17? Something else?

**Comment by polarina** [comment]
Good question. I am unconvinced that promoting a language version request for, say, C++11 to C++17, is a good idea. C++17 brings in a number of incompatibilities with the C++11 standard -- removal of the `register` keyword, for example, which VS will enforce if given `/std:c++17`.

**Comment by nirbheek** [comment]
I think we should emit a warning when an invalid `cpp_std` option is set while building with a version of MSVC that supports the `/std` option. According to [the docs](https://msdn.microsoft.com/en-us/library/mt490614.aspx) the option is supported starting with Visual C++ 2013 Update 3.

**Comment by mikkelfj** [comment]
@polarina 
>The patches should be ready. I am unsure as to why the VS2015 build on AppVeyor is failing. If someone more knowledgeful about AppVeyour could take a look, that'd be appreciated.

I'm not sure at all this is related, but I was having problems with VS2015 x86 for C:

https://github.com/mesonbuild/meson/issues/2134

**Comment by polarina** [comment]
I've rebased the changes on top of current master. Can we get another round of reviews, please?

File: docs/markdown/Release-notes-for-0.46.0.md [hunk 3]
```diff
@@ -21,3 +21,8 @@ Added the function `subdir_done()`. Its invocation exits the current script at
 the point of invocation. All previously invoked build targets and commands are
 build/executed. All following ones are ignored. If the current script was
 invoked via `subdir()` the parent script continues normally.
+
+## Added `cpp_std` option for the Visual Studio C++ compiler
+
+Allows the use of C++17 features and experimental not-yet-standardized
+features. Valid options are `c++11`, `c++14`, `c++17`, and `c++latest`.
```
Comment (nirbheek) [inline docs/markdown/Release-notes-for-0.46.0.md]: This should be added into the `docs/markdown/snippets` directory.

File: mesonbuild/compilers/cpp.py [hunk 4]
```diff
@@ -221,6 +231,9 @@ def get_option_compile_args(self, options):
         std = options['cpp_eh']
         if std.value != 'none':
             args.append('/EH' + std.value)
+        std = options['cpp_std']
+        if std.value not in ['none', 'c++11']:
+            args.append('/std:' + std.value)
```
Comment (nirbheek) [inline mesonbuild/compilers/cpp.py]: Why isn't this value added for `c++11`? Please add a comment about that.

Also, why aren't we warning when `cpp_std:` on an MSVC compiler that does not support the `/std:` argument?

File: test cases/windows/14 vs c++17/meson.build [hunk 5]
```diff
@@ -0,0 +1,8 @@
+project('wincpp', 'cpp')
+
+if meson.get_compiler('cpp').version().version_compare('<19.11')
+    error('MESON_SKIP_TEST Visual Studio 2017 (version 15.3) or later is required for C++17 support')
```
Comment (nirbheek) [inline test cases/windows/14 vs c++17/meson.build]: Why are you skipping the test here? Also where's the `cpp_std:` option set for this test?

**Comment by jpakkane** [comment]
The code around language version has changed so much that this probably requires a full reworking. Thus closing to reduce clutter on the MR page. Please open a new MR with updated code if you get this working against latest trunk. Thanks and sorry for the delay.

**Pull request closed without merging**